etcdutl/snapshot/v3_snapshot.go

```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
 	return nil
 }
 
+// unsafeGetLatestRevision returns the latest revision without locking the given transaction.
+func (s *v3Manager) unsafeGetLatestRevision(tx backend.BatchTx) (latest revision, err error) {
+	err = tx.UnsafeForEach(schema.Key, func(k, _ []byte) (err error) {
+		rev := bytesToRev(k)
+
+		if latest.main < rev.main {
+			latest = rev
+		}
+
+		return nil
+	})
+
+	return
+}
+
+// bumpRevision increases the revision so that, from the client's perspective, the revision never decreases.
+//
+//	If we don't bump the revision, a client looking for a key at one revision finds that the key is actually at a
+//	smaller revision which breaks the promise that revisions never decrease. The client has no way of knowing that
+//	etcd was restored, it only sees the decrease in revision.
+//
+// To ensure that clients don't have a revision that's higher than the restored revision,
+// 'amount' should be a very large number, >= 1_000_000_000.
+//
+// Compaction should be run after this to further help invalidate watcher caches.
+// (see markLatestRevCompacted)
+func (s *v3Manager) bumpRevision(amount int64) error {
+	be := backend.NewDefaultBackend(s.lg, s.outDbPath())
+	defer be.Close()
+
+	tx := be.BatchTx()
+	tx.LockOutsideApply()
+	defer tx.Unlock()
+
+	latest, err := s.unsafeGetLatestRevision(tx)
+	if err != nil {
+		return err
+	}
+
+	s.lg.Info(
+		"bumping latest revision to ensure revisions never decrease",
+		zap.Int64("latest-revision", latest.main),
+		zap.Int64("bump-amount", amount),
+		zap.Int64("new-latest-revision", latest.main+amount),
+	)
+
+	latest.main += amount
+	k := [17]byte{}
+	revToBytes(k[:], latest)
+	tx.UnsafePut(schema.Key, k[:], []byte{})
+
+	return nil
+}
+
+// markLatestRevCompacted sets the scheduled compaction to the latest revision in order to tell the server
+// that it should compact the restored database up to the latest revision.
+//
+// This will cause any watcher to be cancelled as the revision they're watching will be compacted,
```

Comment:
This is restore feature documentation, not a good much for a low level function.